promql/engine.go

```diff
@@ -1104,58 +1072,125 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, storage.Warnings) {
 		enh := &EvalNodeHelper{Out: make(Vector, 0, 1)}
 		// Process all the calls for one time series at a time.
 		it := storage.NewBuffer(selRange)
-		for i, s := range selVS.Series {
-			ev.currentSamples -= len(points)
-			points = points[:0]
-			it.Reset(s.Iterator())
-			ss := Series{
-				// For all range vector functions, the only change to the
-				// output labels is dropping the metric name so just do
-				// it once here.
-				Metric: dropMetricName(selVS.Series[i].Labels()),
-				Points: getPointSlice(numSteps),
-			}
-			inMatrix[0].Metric = selVS.Series[i].Labels()
-			for ts, step := ev.startTimestamp, -1; ts <= ev.endTimestamp; ts += ev.interval {
-				step++
-				// Set the non-matrix arguments.
-				// They are scalar, so it is safe to use the step number
-				// when looking up the argument, as there will be no gaps.
-				for j := range e.Args {
-					if j != matrixArgIndex {
-						otherInArgs[j][0].V = otherArgs[j][0].Points[step].V
-					}
+		// Loop through the unexpanded SeriesSet if Series is nil,
```

Comment:
A lot of tricky code is being duplicated across the branches of this if, you should factor things so this isn't required.

Reply:
Sorry for the late reply, but I have been thinking of a nice way to refactor this to keep the code DRY, but I can't think of a nice way to do it. Thinking of moving the common logic to a function, but since the function would need to be executed every iteration, the function would end up having a lot of parameters, and I don't think that would be nice. Any suggestions here?

Reply:
I think the thing to do here is to change the interface, and push that logic down to the thing providing the series rather than up here in the various consumers. Probably using an iterator.